Make KNOWN_MULTI_FIELDS configurable and add doc

sirosen opened the pull request:
Rather than a module-level constant, move ``KNOWN_MULTI_FIELDS`` to be a parser attribute. This will allow users to override or modify it. However, moving it creates a problem: whereas previously `is_multiple` was just a function which MultiDictProxy could safely call, now it's a method and requires a Parser instance.
(NB: `is_multiple` actually moved from Parser to MultiDictProxy, but my first draft put it on the parser.)

To handle this, change the relationship between the multidictproxy and core modules. `core` now requires `multidictproxy` and the base `Parser` provides a private method for building a MultiDictProxy object.

MultiDictProxy doesn't depend on parsers (tried that, didn't like it). Instead, a MultiDictProxy can be given a `known_multi_fields` argument when created. The `_makeproxy` helper on the Parser class just makes sure to pass `known_multi_fields=self.KNOWN_MULTI_FIELDS`, and everything falls into place from there. Numerous tweaks to parser classes use `_makeproxy` instead of directly instantiating a MultiDictProxy, but that's the bulk of the changeset.

New docs in a section towards the end of the advanced doc explains how multi field detection works and explains how you can set `is_multiple = True` or add a field class to `Parser.KNOWN_MULTI_FIELDS`.

---

If we like this, I think I'll release v7.1 when it merges. If we don't, let's figure out what kind of doc & interface we want instead.

sirosen committed:
Make KNOWN_MULTI_FIELDS configurable and add doc

Rather than a module-level constant, move ``KNOWN_MULTI_FIELDS`` to be
a parser attribute. This will allow users to override or modify it.
However, moving it creates a problem: whereas previously `is_multiple`
was just a function which MultiDictProxy could safely call, now it's
a method and requires a Parser instance.

To handle this, change the relationship between the multidictproxy and
core modules. `core` now requires `multidictproxy` and the base
`Parser` provides a private method for building a MultiDictProxy
object.

MultiDictProxy doesn't depend on parsers (tried that, didn't like it).
Instead, a MultiDictProxy can be given a `known_multi_fields` argument
when created. The `_makeproxy` helper on the Parser class just makes
sure to pass `known_multi_fields=self.KNOWN_MULTI_FIELDS`, and
everything falls into place from there. Numerous tweaks to parser
classes use `_makeproxy` instead of directly instantiating a
MultiDictProxy, but that's the bulk of the changeset.

New docs in a section towards the end of the advanced doc explains how
multi field detection works and explains how you can set
`is_multiple = True` or add a field class to
`Parser.KNOWN_MULTI_FIELDS`.

lafrech commented on docs/advanced.rst:
```diff
@@ -493,6 +493,91 @@ For example, you might implement JSON PATCH according to `RFC 6902 <https://tool
+
+To implement this behavior, webargs will examine schemas for ``marshmallow.fields.List``
+fields. ``List`` fields get unpacked to list values when data is loaded, and
+other fields do not. This also applies fields which inherit from ``List``.
```
applies *to* fields

lafrech commented on docs/advanced.rst:
```diff
@@ -493,6 +493,91 @@ For example, you might implement JSON PATCH according to `RFC 6902 <https://tool
+    str_instance = fields.String()
+
+
+    class CustomMultiplexingField(fields.Field):
```
Why the instance? Couldn't this inherit from String using super to access String methods?

sirosen replied:
Yes, it could/should!
I had just imitated the implementation of `List` (which has a nested field instance) without thinking too hard.

I'll simplify it.

lafrech reviewed (approved):
LGTM. Thanks.

I've been wondering if this should be a major release.

Overrides in parsers using `MultiDictProxy` (`load_querystring`,...) will still work like they used to, just ignoring the `is_multiple` rework.

Should removing `is_multiple` from `core.__all__` be considered a breaking change?

sirosen commented:
> Should removing `is_multiple` from `core.__all__` be considered a breaking change?

I hadn't thought of it as public, but it's certainly possible that someone is importing `is_multiple` and using it.
Which leads to...

> I've been wondering if this should be a major release.

If so (mostly, if we're concerned that removing `is_multiple` could break someone), but you're happy with it as-is, let's mark it as part of 8.0 .
The existing work in the dev branch (changing `is_multiple` behavior to respect the `is_multiple` attribute) could still go out as 7.1 , so we should refrain from merging.

Would you like me to release 7.1 with just what we have today? I'm happy to do so.

lafrech commented:
I don't mind if you ship 7.1 with current dev, then this PR later in 8.0.

But the whole `is_multiple` change (dev + this PR) could be shipped as 8.0. I mean it is simpler to make the whole thing a single version jump, be it major, with all the changes at once without having to document intermediate steps. The `Tuple` change would be part of it as well.

It's not like someone really needs this feature but can't move to 8.0. The only change with 8.0 is related to the feature anyway.

sirosen commented:
Yeah, good point. Why don't we commit to `dev` being the working branch for 8.0 now? That way it's simpler for us and we get to just document the `is_multiple` change for 8.0.

I'm doing cleanup right now to this branch; hopefully done in a few minutes.

sirosen committed:
Improve advanced doc per review

Typo fix + simplify CustomMultiplexingField . Because this is actually
copied from the testsuite, update the tests as well, so that we're
exercising the example code and we know it works.

lafrech commented:
> Why don't we commit to dev being the working branch for 8.0 now? That way it's simpler for us and we get to just document the is_multiple change for 8.0.

Exactly.

PR approved.
